keps/sig-storage/3756-volume-reconstruction/README.md

```diff
@@ -0,0 +1,905 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+### Goals
+
+* During kubelet startup, allow it to populate additional information about
+  _how_ are existing volumes mounted.
+  [KEP 1710](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling)
+  needs to know what mount options did the previous kubelet used when mounting
+  the volumes, to be able to tell if they need any change or not.
+* Fix [#105536](https://github.com/kubernetes/kubernetes/issues/105536): Volumes
+  are not cleaned up (unmounted) after kubelet restart, which needs a similar
+  VolumeManager refactoring.
+* In general, make volume cleanup more robust.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+ 
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Introduction
+
+*VolumeManager* is a piece of kubelet that mounts volumes that should be
+mounted (i.e. a Pod that needs the volume exists) and unmounts volumes that are
+not needed any longer (all Pods that used them were deleted).
+
+VolumeManager keeps two caches:
+* *DesiredStateOfWorld* (DSW) contains volumes that should be mounted.
+* *ActualStateOfWorld* (ASW) contains currently mounted volumes.
+
+VolumeManager then compares these two caches and tries to move ASW towards DSW.
+
+Both caches exist only in memory and are lost when kubelet process dies. It's
+relatively easy to populate DSW - just list all Pods from the API server
+(and static pods and whatnot) and collect their volumes.
+
+*Volume reconstruction* is a process where kubelet tries to create a single
+valid `PersistentVolumeSpec` or `VolumeSpec` for a volume from the OS.
+Typically from mount table by looking at what's mounted at 
+`/var/lib/kubelet/pods/*/volumes/XYZ`. This process is imperfect,
+it populates only `(Persistent)VolumeSpec` fields that are necessary to unmount
+the volume (i.e. to call `volumePlugin.TearDown` + `UnmountDevice` calls).
+
+Today, kubelet populates VolumeManager's DSW first, from static Pods and pods
+received from the API server. ASW is populated from the OS
+after DSW is complete and **only volumes missing in DSW are added there**.
+In other words, kubelet reconstructs only the volumes for Pods that were running
+before kubelet started, but they were delered when kubelet started. (If the pod
+was Running, its volumes would be in DSW).
+
+We assumed that this was enough, because if a volume is in DSW, the
+VolumeManager will try to mount the volume, and it will eventually reach ASW. 
+
+We needed to add
+[a complex workaround](https://github.com/kubernetes/kubernetes/pull/110670)
+to actually unmount a volume if it's initially in DSW, but user deletes all
+Pods that need it before the volume reaches ASW. 
```

Comment:
Since this feature directly goes beta, we will have to test if this new mechanism fixes the bug PR#110670 fixes or at least co-exists without stomping on each others toes.

Reply:
Do we have a e2e / integration test for #110670 ?

Reply:
No I don't think I wrote any e2e for this one. We only have unit tests.

Reply:
We still have the unit test, I think.